File: examples/dockq_demo.rs
//! DockQ interface quality assessment demo.
//!
//! Demonstrates DockQ scoring for protein-protein complex quality evaluation.
//!
//! Run with:
//! ```
//! cargo run --example dockq_demo --features dockq
//! ```

use pdbrust::PdbStructure;
use pdbrust::dockq::{ChainMappingStrategy, DockQOptions};
use pdbrust::records::Atom;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== DockQ Interface Quality Assessment Demo ===\n");

    // Create a synthetic two-chain complex (native)
    let native = create_complex();
    println!(
        "Native structure: {} atoms, chains: {:?}",
        native.atoms.len(),
        native.get_chain_ids()
    );

    // ========================================================================
    // 1. Self-comparison (perfect score)
    // ========================================================================
    println!("\n--- 1. Self-comparison ---");
    let result = native.dockq_to(&native)?;
    print_result(&result);

    // ========================================================================
    // 2. Slightly perturbed model
    // ========================================================================
    println!("\n--- 2. Slightly perturbed model (chain B shifted 1 A) ---");
    let mut perturbed = native.clone();
    for atom in &mut perturbed.atoms {
        if atom.chain_id == "B" {
            atom.y += 1.0;
        }
    }

    let options = DockQOptions {
        chain_mapping: ChainMappingStrategy::Explicit(vec![
            ("A".to_string(), "A".to_string()),
            ("B".to_string(), "B".to_string()),
        ]),
        ..Default::default()
    };
    let result = perturbed.dockq_to_with_options(&native, options)?;
    print_result(&result);

    // ========================================================================
    // 3. Badly docked model
    // ========================================================================
    println!("\n--- 3. Badly docked model (chain B shifted 15 A) ---");
    let mut bad_model = native.clone();
    for atom in &mut bad_model.atoms {
        if atom.chain_id == "B" {
            atom.y += 15.0;
        }
    }

    let options = DockQOptions {
        chain_mapping: ChainMappingStrategy::Explicit(vec![
            ("A".to_string(), "A".to_string()),
            ("B".to_string(), "B".to_string()),
        ]),
        ..Default::default()
    };
    let result = bad_model.dockq_to_with_options(&native, options)?;
    print_result(&result);

    // ========================================================================
    // 4. Auto chain mapping with swapped labels
    // ========================================================================
    println!("\n--- 4. Auto chain mapping (model has swapped chain labels) ---");
    let mut swapped = PdbStructure::new();
    for atom in &native.atoms {
        let mut new_atom = atom.clone();
        new_atom.chain_id = if atom.chain_id == "A" {
            "B".to_string()
        } else {
            "A".to_string()
        };
        swapped.atoms.push(new_atom);
    }

    let result = swapped.dockq_to(&native)?;
    println!("Chain mapping: {:?}", result.chain_mapping);
    print_result(&result);

    println!("\n=== Demo complete ===");
    Ok(())
}

fn print_result(result: &pdbrust::dockq::DockQResult) {
    println!(
        "  Total DockQ: {:.4} ({} interface{})",
        result.total_dockq,
        result.num_interfaces,
        if result.num_interfaces != 1 { "s" } else { "" }
    );

    for iface in &result.interfaces {
        println!(
            "  Interface {}-{} (model {}-{}):",
            iface.native_chains.0,
            iface.native_chains.1,
            iface.model_chains.0,
            iface.model_chains.1,
        );
        println!("    DockQ:  {:.4} ({})", iface.dockq, iface.quality);
        println!("    fnat:   {:.4}", iface.fnat);
        println!("    iRMSD:  {:.4} A", iface.irmsd);
        println!("    LRMSD:  {:.4} A", iface.lrmsd);
        println!("    F1:     {:.4}", iface.f1);
        println!(
            "    Contacts: {} native, {} model",
            iface.num_native_contacts, iface.num_model_contacts
        );
    }
}

fn create_complex() -> PdbStructure {
    let mut structure = PdbStructure::new();
    let mut serial = 1;

    let chain_a_residues = ["ALA", "GLY", "VAL", "LEU", "ILE", "PHE", "TRP"];
    let chain_b_residues = ["SER", "THR", "ASP", "GLU", "LYS", "ARG", "HIS"];

    // Chain A along x-axis at y=0
    for (i, resname) in chain_a_residues.iter().enumerate() {
        let x = i as f64 * 3.8;
        let seq = (i + 1) as i32;
        for &(name, dx, dy, element) in &[
            ("N", -0.5, -0.5, "N"),
            ("CA", 0.0, 0.0, "C"),
            ("C", 0.5, 0.0, "C"),
            ("O", 0.5, 0.5, "O"),
        ] {
            structure.atoms.push(Atom {
                serial,
                name: name.to_string(),
                alt_loc: None,
                residue_name: resname.to_string(),
                chain_id: "A".to_string(),
                residue_seq: seq,
                ins_code: None,
                is_hetatm: false,
                x: x + dx,
                y: dy,
                z: 0.0,
                occupancy: 1.0,
                temp_factor: 20.0,
                element: element.to_string(),
            });
            serial += 1;
        }
    }

    // Chain B along x-axis at y=4.0 (within contact distance)
    for (i, resname) in chain_b_residues.iter().enumerate() {
        let x = i as f64 * 3.8;
        let seq = (i + 1) as i32;
        for &(name, dx, dy, element) in &[
            ("N", -0.5, 3.5, "N"),
            ("CA", 0.0, 4.0, "C"),
            ("C", 0.5, 4.0, "C"),
            ("O", 0.5, 4.5, "O"),
        ] {
            structure.atoms.push(Atom {
                serial,
                name: name.to_string(),
                alt_loc: None,
                residue_name: resname.to_string(),
                chain_id: "B".to_string(),
                residue_seq: seq,
                ins_code: None,
                is_hetatm: false,
                x: x + dx,
                y: dy,
                z: 0.0,
                occupancy: 1.0,
                temp_factor: 20.0,
                element: element.to_string(),
            });
            serial += 1;
        }
    }

    structure
}

File: pdbrust-python/examples/dockq_demo.py
"""DockQ interface quality assessment demo.

Demonstrates DockQ scoring for evaluating protein-protein docking quality.

Usage:
    cd pdbrust-python
    maturin develop --release
    python examples/dockq_demo.py
"""

import os
import sys

import pdbrust
from pdbrust import (
    ChainMappingStrategy,
    DockQOptions,
    DockQQuality,
)


def main():
    print("=== DockQ Interface Quality Assessment (Python) ===\n")

    # Find the 1HSG PDB file (HIV protease dimer with chains A and B)
    script_dir = os.path.dirname(os.path.abspath(__file__))
    pdb_path = os.path.join(script_dir, "..", "..", "examples", "pdb_files", "1HSG.pdb")

    if not os.path.exists(pdb_path):
        print(f"PDB file not found: {pdb_path}")
        sys.exit(1)

    native = pdbrust.parse_pdb_file(pdb_path)
    print(f"Loaded: {native}")
    print(f"Chains: {native.get_chain_ids()}")
    print(f"Atoms:  {native.num_atoms}")

    # ================================================================
    # 1. Self-comparison (perfect DockQ = 1.0)
    # ================================================================
    print("\n--- 1. Self-comparison (expect DockQ = 1.0) ---")
    result = native.dockq_to(native)
    print_result(result)

    # ================================================================
    # 2. Using explicit chain mapping
    # ================================================================
    print("\n--- 2. Explicit chain mapping ---")
    options = DockQOptions(
        chain_mapping=ChainMappingStrategy.explicit([("A", "A"), ("B", "B")])
    )
    result = native.dockq_to_with_options(native, options)
    print(f"  Chain mapping: {result.chain_mapping}")
    print_result(result)

    # ================================================================
    # 3. Inspecting per-interface results
    # ================================================================
    print("\n--- 3. Per-interface details ---")
    result = native.dockq_to(native)
    for i, iface in enumerate(result.interfaces):
        print(f"  Interface {i + 1}:")
        print(f"    Native chains: {iface.native_receptor_chain}-{iface.native_ligand_chain}")
        print(f"    Model chains:  {iface.model_receptor_chain}-{iface.model_ligand_chain}")
        print(f"    DockQ:  {iface.dockq:.4f}")
        print(f"    fnat:   {iface.fnat:.4f}")
        print(f"    fnonnat:{iface.fnonnat:.4f}")
        print(f"    iRMSD:  {iface.irmsd:.4f} A")
        print(f"    LRMSD:  {iface.lrmsd:.4f} A")
        print(f"    F1:     {iface.f1:.4f}")
        print(f"    Native contacts: {iface.num_native_contacts}")
        print(f"    Model contacts:  {iface.num_model_contacts}")
        print(f"    Clashes: {iface.num_clashes}")

    # ================================================================
    # 4. Quality classification
    # ================================================================
    print("\n--- 4. Quality classification ---")
    for score in [0.10, 0.30, 0.60, 0.90]:
        quality = DockQQuality.from_score(score)
        print(f"  DockQ = {score:.2f} -> {quality}")

    # ================================================================
    # 5. Indexing into result
    # ================================================================
    print("\n--- 5. DockQResult iteration ---")
    result = native.dockq_to(native)
    print(f"  Number of interfaces: {len(result)}")
    if len(result) > 0:
        first = result[0]
        print(f"  First interface: {first}")

    print("\n=== Demo complete ===")


def print_result(result):
    """Print a DockQ result summary."""
    print(f"  Total DockQ: {result.total_dockq:.4f} ({result.num_interfaces} interface(s))")
    for iface in result.interfaces:
        print(
            f"    {iface.native_receptor_chain}-{iface.native_ligand_chain}: "
            f"DockQ={iface.dockq:.4f} fnat={iface.fnat:.4f} "
            f"iRMSD={iface.irmsd:.2f} LRMSD={iface.lrmsd:.2f} "
            f"({iface.quality})"
        )


if __name__ == "__main__":
    main()
